### ISSUE TITLE
[JENKINS-27648] Introduce directory disposal service

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/jenkins#20995]
<p>Jenkins discards directories that are not needed any longer all the time. Aside from the fact it is subject of different random failures (on windows, on NFS, etc.), it is also a time consuming operation.</p>

<p>Instead of deleting such files synchronously, I propose to create Jenkins-wide service where directories can be registered for later disposal. The service will rename the directory so it will appear deleted immediately and delete the renamed one later.</p>

<p>Similar approach is implemented in <a href="https://issues.jenkins-ci.org/browse/JENKINS-20056" class="external-link" target="_blank" rel="nofollow noopener">Workspace cleanup plugin</a> though the solution leaks renamed directories in case the deletion fails.</p>

<p>Such service will track registered directories even after the deletion failed and retry as many times as necessary or possibly report to admin that there is a problem Jenkins itself cannot handle despite repeated attempts.</p>

<p>There are several areas that can benefit from this aside from Workspace cleanup. though, I am not sure if this behaviour should replace what <tt>FilePath#deleteRecursive</tt> currently does.</p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/olivergondza.png" title="olivergondza's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=olivergondza">olivergondza</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-27648" target="_blank">Introduce directory disposal service</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/olivergondza.png" title="olivergondza's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=olivergondza">olivergondza</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Minor
<li><b>component(s)</b>: core
<li><b>resolution</b>: Won't Fix
<li><b>resolved</b>: 2015-04-03T09:50:26+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 1
<li><b>imported</b>: 2025-11-24
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>Jenkins discards directories that are not needed any longer all the time. Aside from the fact it is subject of different random failures (on windows, on NFS, etc.), it is also a time consuming operation.</p>

<p>Instead of deleting such files synchronously, I propose to create Jenkins-wide service where directories can be registered for later disposal. The service will rename the directory so it will appear deleted immediately and delete the renamed one later.</p>

<p>Similar approach is implemented in <a href="https://issues.jenkins-ci.org/browse/JENKINS-20056" class="external-link" target="_blank" rel="nofollow noopener">Workspace cleanup plugin</a> though the solution leaks renamed directories in case the deletion fails.</p>

<p>Such service will track registered directories even after the deletion failed and retry as many times as necessary or possibly report to admin that there is a problem Jenkins itself cannot handle despite repeated attempts.</p>

<p>There are several areas that can benefit from this aside from Workspace cleanup. though, I am not sure if this behaviour should replace what <tt>FilePath#deleteRecursive</tt> currently does.</p></pre>
</details>
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/danielbeck.png" title="danielbeck's avatar" /> danielbeck:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-27648?focusedId=225163&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-225163">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>FWIW this could be interesting if we ever move to a 'Recycle Bin'/delayed deletion model for deleting builds/jobs.</p>
</pre>
</details>

</ul>
</details>

<p>FWIW this could be interesting if we ever move to a 'Recycle Bin'/delayed deletion model for deleting builds/jobs.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/olivergondza.png" title="olivergondza's avatar" /> olivergondza:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-27648?focusedId=225170&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-225170">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=danielbeck" class="user-hover" rel="danielbeck">danielbeck</a>, I have seen UI threads timeout deleting jobs while waiting on FS. Which is something my PR would address.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=danielbeck" class="user-hover" rel="danielbeck">danielbeck</a>, I have seen UI threads timeout deleting jobs while waiting on FS. Which is something my PR would address.</p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `relates_to` from Jira: <a href="https://github.com/jenkinsci/jenkins/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-24824%22">JENKINS-24824</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `relates_to` from Jira: <a href="https://github.com/jenkinsci/jenkins/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-53579%22">JENKINS-53579</a>]</i>
